Avoid Coordinator#getClusterFormationState() blocking on Coordinator#mutex

joshua-adams-1 opened the pull request:
Modifies `Coordinator#getClusterFormationState()` to not block on `getLastAcceptedTerm()` and `getCurrentTerm()`, each of which block on the mutex which may be busy doing other intensive work. A `ClusterFormationClusterStateView` is introduced which stores a subset of the cluster state for use in the health reporting. This is an unsynchronised update of the cluster state and current term, updated under the `mutex` but not read while under it, so it may be outdated. However, it unblocks health reporting when the coordinator may be busy elsewhere

Closes: 135947
Closes: ES-13110

joshua-adams-1 committed:
Avoid Coordinator#getClusterFormationState() blocking on
Coordinator#mutex

Modifies Coordinator#getClusterFormationState() to not block on
getLastAcceptedTerm() and getCurrentTerm(), each of which block on the
mutex which may be busy doing other intensive work. An unsynchronised
cluster state and current term are stored in the Coordinator, updated
beneath the mutex but not read while under it, so it may be outdated.
However, it unblocks health reporting when the coordinator may be busy
elsewhere

Closes: 135947
Relates to: ES-13110

DaveCTurner commented:
I might suggest doing it all at once tbh, we ultimately don't want two separate `volatile` fields here.

Also in terms of testing, I think it'd be enough to add relevant checks to `Coordinator#invariant()` saying that these values match the properly-synchronized ones while under the `mutex`.

joshua-adams-1 committed:
Avoid Coordinator#getClusterFormationState() blocking on
Coordinator#mutex

Modifies Coordinator#getClusterFormationState() to not block on
getLastAcceptedTerm() and getCurrentTerm(), each of which block on the
mutex which may be busy doing other intensive work. A
ClusterFormationClusterStateView is introduced which stores a subset of
the cluster state for use in the health reporting. This is an
unsynchronised update of the cluster state and current term, updated
under the mutex but not read while under it, so it may be outdated.
However, it unblocks health reporting when the coordinator may be busy
elsewhere

Closes: 135947
Relates to: ES-13110

joshua-adams-1 committed:
Merge branch 'main' into cluster-formation-state-constructor-without-cluster-state-param

joshua-adams-1 committed:
Add comment to updateClusterFormationClusterStateView

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java:
```diff
@@ -177,36 +190,35 @@ public record ClusterFormationState(
+                clusterFormationClusterStateView.lastAcceptedVersion,
+                clusterFormationClusterStateView.lastAcceptedTerm,
+                clusterFormationClusterStateView.lastAcceptedConfiguration,
+                clusterFormationClusterStateView.lastCommittedConfiguration,
```
Maybe just pass in `clusterFormationClusterStateView` itself, no real need to deconstruct it like this I reckon.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -335,6 +341,15 @@ public Coordinator(
+            0,
+            0,
+            null,
+            null,
```
Could we use an empty map and `org.elasticsearch.cluster.coordination.CoordinationMetadata.VotingConfiguration#EMPTY_CONFIG` rather than nulls here. Bit of a pain we can't fill in the local disco node too but we don't know it at this point in startup.

DaveCTurner reviewed (commented):
I like it :)

joshua-adams-1 committed:
Do not deconstruct clusterFormationClusterStateView when passing into ClusterFormationState

joshua-adams-1 committed:
Use record equality in invariant()

joshua-adams-1 committed:
Merge branch 'main' into cluster-formation-state-constructor-without-cluster-state-param

joshua-adams-1 committed:
Fix mutateResponse

joshua-adams-1 committed:
Merge branch 'cluster-formation-state-constructor-without-cluster-state-param' of https://github.com/joshua-adams-1/elasticsearch into cluster-formation-state-constructor-without-cluster-state-param

joshua-adams-1 committed:
Fix calculateHasDiscoveredQuorum parameters

Commit:
[CI] Auto commit changes from spotless

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java:
```diff
@@ -216,39 +242,28 @@ public ClusterFormationState(
                 in.readCollectionAsImmutableList(TransportAddress::new),
                 in.readCollectionAsImmutableList(DiscoveryNode::new),
                 in.readCollectionAsImmutableSet(DiscoveryNode::new),
-                in.readLong(),
```
Nice idea, but this reorders the fields as sent on the wire so it won't work.

If we really want to do this we'd need to introduce a new `TransportVersion` for the new order, and write an adapter to shuffle the fields around when reading from an older node.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -1239,6 +1277,21 @@ public void invariant() {
+                    lastAcceptedClusterState.getLastAcceptedConfiguration(),
+                    lastAcceptedClusterState.getLastCommittedConfiguration(),
+                    getCurrentTerm()
+                );
```
Can we just `new ClusterFormationFailureHelper.ClusterFormationClusterStateView(lastAcceptedClusterState, getCurrentTerm())` here?

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -344,17 +360,31 @@ public Coordinator(
+     * {@code clusterFormationClusterStateView} record too.
+     */
+    private void updateClusterFormationClusterStateView() {
+        ClusterState currentClusterState = coordinationState.get().getLastAcceptedState();
```
nit: either `final` or else just inline, no need for a separate variable really

DaveCTurner reviewed (commented):
I like it, apart from the change to the wire format that needs further work (or reverting).

Could we also have a test that:

1. starts a background thread which acquires `Coordinator#mutex` and then blocks on a `safeAwait()` of some kind
2. calls `Coordinator#getClusterFormationState()`
3. releases whatever is blocking the background thread

That way we can be sure that nothing needs that mutex.

joshua-adams-1 committed:
Comments

joshua-adams-1 committed:
Merge branch 'cluster-formation-state-constructor-without-cluster-state-param' of https://github.com/joshua-adams-1/elasticsearch into cluster-formation-state-constructor-without-cluster-state-param

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:
```diff
@@ -414,6 +443,9 @@ private void handleApplyCommit(ApplyCommitRequest applyCommitRequest, ActionList
             logger.trace("handleApplyCommit: applying commit {}", applyCommitRequest);
 
             coordinationState.get().handleCommit(applyCommitRequest);
+            // Since the coordination state changed, we update accordingly
```
These comments seem unnecessary (and indeed positively distracting) to me - really it's not saying much more than "the next line calls `updateClusterFormationClusterStateView()` because it has to" but such comments could be applied to basically every line of code here.

joshua-adams-1 committed:
Fix serialization error

joshua-adams-1 committed:
Add unit test for mutex

joshua-adams-1 committed:
Merge branch 'main' into cluster-formation-state-constructor-without-cluster-state-param

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java:
```diff
@@ -2091,6 +2092,45 @@ protected long transportDelayMillis(String actionName) {
+
+            // Latches for coordination
+            final var mutexAcquired = new CountDownLatch(1);
+            final var release = new CountDownLatch(1);
```
Consider using a `new CyclicBarrier(2)` for this kind of release-and-wait/wait-and-release pattern, it's a little shorter.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java:
```diff
@@ -2091,6 +2092,45 @@ protected long transportDelayMillis(String actionName) {
+                        release.await();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
```
Definitely shouldn't be waiting forever here, and should fail on an interrupt:

```suggestion
                    mutexAcquired.countDown();
                    safeAwait(release);
```

But also this could just be

```suggestion
                    safeAwait(barrier);
                    safeAwait(barrier);
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java:
```diff
@@ -2091,6 +2092,45 @@ protected long transportDelayMillis(String actionName) {
+            t.start();
+
+            // Wait for the mutex to be acquired
+            assertTrue(mutexAcquired.await(10, java.util.concurrent.TimeUnit.SECONDS));
```
```suggestion
            safeAwait(mutexAcquired);
```

or even

```suggestion
            safeAwait(barrier);
```

DaveCTurner reviewed (commented):
Production code LGTM, couple of tips/suggestions on the new test

joshua-adams-1 committed:
Fix testGetClusterFormationStateDoesNotBlockOnMutex

joshua-adams-1 committed:
Nits

joshua-adams-1 committed:
Merge branch 'cluster-formation-state-constructor-without-cluster-state-param' of https://github.com/joshua-adams-1/elasticsearch into cluster-formation-state-constructor-without-cluster-state-param

joshua-adams-1 commented on server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java:
```diff
@@ -2091,6 +2092,34 @@ protected long transportDelayMillis(String actionName) {
 
+    public void testGetClusterFormationStateDoesNotBlockOnMutex() {
+        try (Cluster cluster = new Cluster(1)) {
+            cluster.stabilise();
```
Good catch!

joshua-adams-1 committed:
Apply suggestions from code review

Co-authored-by: David Turner <david.turner@elastic.co>

joshua-adams-1 committed:
Change to cyclic barrier

joshua-adams-1 committed:
Merge branch 'cluster-formation-state-constructor-without-cluster-state-param' of https://github.com/joshua-adams-1/elasticsearch into cluster-formation-state-constructor-without-cluster-state-param

joshua-adams-1 committed:
Merge branch 'main' of github.com:elastic/elasticsearch into cluster-formation-state-constructor-without-cluster-state-param

joshua-adams-1 committed:
Fix serialization

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java:
```diff
@@ -2091,6 +2092,29 @@ protected long transportDelayMillis(String actionName) {
+    public void testGetClusterFormationStateDoesNotBlockOnMutex() {
+        try (Cluster cluster = new Cluster(1)) {
+            cluster.runRandomly();
+            cluster.stabilise();
```
still think we should remove this `stabilise()`

joshua-adams-1 replied:
Is there redundancy in keeping it? I was copying a pattern found in other tests in the suite, such as `testPeerFinderListener` [here](https://github.com/elastic/elasticsearch/blob/7ababc02bfa58ab18f25d94864484084a7b9bf9f/server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java#L1948).

DaveCTurner replied:
It weakens the test to one that verifies we only don't require the mutex when the cluster is fully stabilized, whereas in fact we want to assert that we _never_ require the mutex - indeed we effectively only actually call this method while the cluster is not fully-stabilized.

Most of these tests do want to start with the cluster in a fully-stabilized state. Arguably `testPeerFinderListener` should also be strengthened similarly.

joshua-adams-1 replied:
That makes sense to me! I shall remove

DaveCTurner reviewed (approved):
LGTM (one nit)

joshua-adams-1 committed:
Remove stabilisation check

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java:
```diff
@@ -177,37 +225,27 @@ public record ClusterFormationState(
-                clusterState.term(),
-                clusterState.getLastAcceptedConfiguration(),
-                clusterState.getLastCommittedConfiguration(),
+                clusterFormationClusterStateView.localNode,
```
Ackd, I will open a follow-up PR to show what I mean here (or prove to myself that this is better in the course of preparing such a PR 😁 )

joshua-adams-1 committed:
Merge branch 'main' into cluster-formation-state-constructor-without-cluster-state-param
